### ISSUE TITLE
session 5. back-propagation

### ISSUE THREAD

**Issue by Odin-son** [Odin-son/felis-catus#4]

**Comment by Odin-son** [comment]
신경망의 순전파에서 수행하는 행렬의 곱은 기하학에서는 affine transform 이라고 함
affine transform process = "affine 계층"

**Comment by Odin-son** [comment]
Input -> Affine -> Relu -> ... -> Affine-> Relu -> Affine -> Softmax -> (확률)

Softmax 계층은 정규화(합이 1)하여 출력

**Comment by Odin-son** [comment]
* 신경망 - 학습 & 추론
   * 학습 : ~ Softmax
   * 추론 : 일반적으로 Affine 계층을 출력 결과로 사용함, 이를 점수(score)라 함, 답을 하나만 내는 경우엔 가장 높은 점수만 알면 되니까.

**Comment by Odin-son** [comment]
Softmax 계층은 손실함수인 CEE 까지 포함하여 Softmax-with-Loss 계층을 구현.

**Comment by Odin-son** [comment]
신경망 학습의 목적은 신경망의 출력(Softmax의 출력)이 정답 레이블과 가까워지도록 가중치의 매개변수의 값을 조정하는 것

**Comment by Odin-son** [comment]
수치 미분과 backpropagation 의 차이는 구현상의 장단점에 있다. 수치 미분은 쉬운 반면 속도가 느리고 backpropagation은 속도가 빠르지만, 복잡해서 실수하기 마련이다.

따라서 backpropagation으로 구현한 사항을 수치 미분의 결과 값과 같은지 검증을 하는 과정이 필요하다

**Comment by Odin-son** [comment]
softmax 함수에서 차원을 생각해줘야되는 이유는 np.max가 전체 행렬의 최대값을 찾기때문에
행마다 해줄수가 없어서 Transpose 후 np.max(x, axis=0) 로 그 값을 구한다.

**Comment by Odin-son** [comment]
>>> np.max(dY)
6
>>> np.max(dY.T, axis = 0)
array([3, 6])
